tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -66,6 +70,28 @@ private[scalajs] final class ScalaJSClassEmitter(
 
   private[emitter] def semantics: Semantics = linkingUnit.semantics
 
+  private[emitter] def genBracketSelect(qual: js.Tree, item: js.Tree)(
+      implicit pos: Position): js.Tree = {
+    item match {
+      case js.StringLiteral(name) if internalOptions.optimizeBracketSelects &&
+          isValidIdentifier(name) && name != "eval" =>
+        /* We exclude "eval" because Rhino does not respect the strict mode
+         * specificities of eval().
+         */
+        js.DotSelect(qual, js.Ident(name))
+      case _ =>
+        js.BracketSelect(qual, item)
+    }
+  }
+
+  private[emitter] def genIdentBracketSelect(qual: js.Tree, item: String)(
+      implicit pos: Position): js.Tree = {
+    if (internalOptions.optimizeBracketSelects)
+      js.DotSelect(qual, js.Ident(item))
+    else
+      js.BracketSelect(qual, js.StringLiteral(item))
+  }
+
```

Comment:
Actually, I don't understand why these methods are in `ScalaJSClassEmitter`. They have nothing to do with class emission specifically and it shows: The fact that you need to pass a `ScalaJSClassEmitter` to a `JSDesugaring` to generate constructors (for merely calling back the `emitter` again...).